source/adaptive_load/metrics_plugin_impl.cc

```diff
@@ -0,0 +1,85 @@
+#include "adaptive_load/metrics_plugin_impl.h"
+
+#include "envoy/registry/registry.h"
+
+#include "external/envoy/source/common/common/assert.h"
+#include "external/envoy/source/common/protobuf/protobuf.h"
+
+#include "absl/strings/numbers.h"
+
+namespace Nighthawk {
+
+NighthawkStatsEmulatedMetricsPlugin::NighthawkStatsEmulatedMetricsPlugin(
+    const nighthawk::client::Output& nighthawk_output) {
+  for (const nighthawk::client::Result& result : nighthawk_output.results()) {
+    if (result.name() != "global") {
+      continue;
+    }
+    int concurrency;
+    // "auto" is not supported since we have no way to detect the number of cores Nighthawk Service
+    // used; "auto" should have been caught during input validation.
+    RELEASE_ASSERT(absl::SimpleAtoi(nighthawk_output.options().concurrency().value(), &concurrency),
+                   "Concurrency 'auto' not supported in Adaptive Load Controller.");
+    const long total_specified = nighthawk_output.options().requests_per_second().value() *
+                                 nighthawk_output.options().duration().seconds() * concurrency;
+    int total_sent = 0;
+    int total_2xx = 0;
+    for (const nighthawk::client::Counter& counter : result.counters()) {
+      if (counter.name() == "benchmark.http_2xx") {
+        total_2xx = counter.value();
+      } else if (counter.name() == "upstream_rq_total") {
+        total_sent = counter.value();
+      }
+    }
+    metric_from_name_["attempted-rps"] =
+        static_cast<double>(total_specified) /
+        static_cast<double>(nighthawk_output.options().duration().seconds());
+    metric_from_name_["achieved-rps"] =
+        static_cast<double>(total_sent) /
+        static_cast<double>(nighthawk_output.options().duration().seconds());
+    metric_from_name_["send-rate"] =
+        static_cast<double>(total_sent) / static_cast<double>(total_specified);
+    metric_from_name_["success-rate"] =
+        static_cast<double>(total_2xx) / static_cast<double>(total_sent);
+
+    for (const nighthawk::client::Statistic& statistic : result.statistics()) {
+      if (statistic.id() == "benchmark_http_client.request_to_response") {
+        double min = statistic.min().seconds() * 1.0e9 + statistic.min().nanos();
+        double mean = statistic.mean().seconds() * 1.0e9 + statistic.mean().nanos();
```

Comment:
We have the `const long kOneBillion = 1000 * 1000 * 1000;`, which is nice. Consider moving that to a place where we can share it, and re-use that here for consistency? I think existing places in the code bases could benefit from this too; out of scope here, but might be worth filing an issue to suggest a pass to clean that up afterwards.

Reply:
I ended up finding some Google time utilities for dealing with Durations -- filed #434 to apply it elsewhere.